src/services/DomainModel.ts

```diff
@@ -0,0 +1,48 @@
+// Do not assume you can reach every node, sensor, or reading starting from a
+// gateway you're looking at. Be mindful that a Query is probably in effect
+// which has limited the extent to which we have populated all the objects into
+// memory. For example, if you're working with recent data from the  'Black
+// Pearl' node you might have info about the Gateway, but just the one
+// node, and a few days of readings. You'll find `Unknown`s at the points in the
+// tree that we don't care about in the context of an active Query.
+
+export interface Unknown {
+  // Unknown means you've reached an unexplored part of the tree, not a dead end.
+  explored: false;
+}
+
+export interface Query {
+  params: QueryParam[]; // TBD what a QueryParam is
+  results: Gateway[] | Node[] | Sensor[] | Reading[];
+}
+
+export interface Gateway {
+  gatewayID: string;
+  nodes: Node[] | Unknown;
+}
+
+export interface Node {
+  nodeID: string;
+  gateway: Gateway | null | Unknown;
+  sensors: Sensor[] | Unknown;
+  name: string | Unknown;
+  location: string | Unknown;
+  description: string | Unknown;
+}
+
+export interface Sensor {
+  node: Node;
+  schema: ReadingSchema;
+  readings: Reading[] | Unknown;
+}
+
+export interface ReadingSchema {
+  unit: string; // e.g. Kelvin
```

Comment:
It’s still not 100% clear to me how we’ll avoid having at least one `enum` or `switch` for handling these units. If I modeling this naively I’d be inclined to have an `enum` for `unit`.

But I know you all have plans for this, so I’ll remain in wait-and-see mode. Just sharing my thoughts on this one.

Reply:
The unit is for display and definition purposes so that a "generic" implementation of a sensor renderer can simply take the value and the unit and put them together to produce meaningful output. The unit could also be used to label the y axis on graphs. 

Having said that, I imagine each sensor type will have its own custom renderer too so rendering can be more specific.